Firestore/core/src/firebase/firestore/model/mutation_batch_result.cc

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/mutation_batch_result.h"
+
+#include <ostream>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/model/mutation.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation_batch.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/to_string.h"
+#include "absl/strings/str_cat.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+MutationBatchResult::MutationBatchResult(
+    MutationBatch batch,
+    model::SnapshotVersion commit_version,
+    std::vector<model::MutationResult> mutation_results,
+    absl::optional<nanopb::ByteString> stream_token)
+    : batch_(std::move(batch)),
+      commit_version_(commit_version),
+      mutation_results_(std::move(mutation_results)),
+      stream_token_(std::move(stream_token)) {
+  HARD_ASSERT(batch_.mutations().size() == mutation_results_.size(),
+              "Mutations sent %s must equal results received %s",
+              batch_.mutations().size(), mutation_results_.size());
+
+  const auto& mutations = batch_.mutations();
+  for (size_t i = 0; i < mutations.size(); i++) {
+    absl::optional<SnapshotVersion> version = mutation_results_[i].version();
+    if (!version) {
+      // deletes don't have a version, so we substitute the commit_version
+      // of the entire batch.
+      version = commit_version_;
+    }
+
+    doc_versions_[mutations[i].key()] = *version;
```

Comment:
Optional: create an intermediate variable for the `key` to avoid two different nested pairs of braces?

Reply:
Done.